### PR TITLE
[Validator] Add AtLeastOne constraint and validator

### PR DESCRIPTION
**Pull request by przemyslaw-bogusz**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT
| Doc PR        | TODO

This constraint allows you to apply a collection of constraints to a value, and it will be considered valid, if it satisfies at least one of the constraints from the collection.

Some examples:

```php
    /**
     * @Assert\AtLeastOne({
     *     @Assert\Length(min=5),
     *     @Assert\EqualTo("bar")
     * })
     */
    public $name = 'foo';

    /**
     * @Assert\AtLeastOne({
     *     @Assert\All({@Assert\GreaterThanOrEqual(10)}),
     *     @Assert\Count(20)
     * })
     */
    public $numbers = ['3', '5'];

    /**
     * @Assert\All({
     *     @Assert\AtLeastOne({
     *          @Assert\GreaterThanOrEqual(5),
     *          @Assert\LessThanOrEqual(3)
     *     })
     * })
     */
    public $otherNumbers = ['4', '5'];
```

The respective default messages would be:
`name: This value should satisfy at least one of the following constraints: [1] This value is too short. It should have 5 characters or more. [2] This value should be equal to "bar".`

`numbers: This value should satisfy at least one of the following constraints: [1] Each element of this collection should satisfy its own set of constraints. [2] This collection should contain exactly 20 elements.`

`otherNumbers[0]: This value should satisfy at least one of the following constraints: [1] This value should be greater than or equal to 5. [2] This value should be less than or equal to 3.`

But of course you could also create a simple custom message like `None of the constraints are satisfied`.

### CODE REVIEW

**Review by nicolas-grekas — COMMENTED** [review]
Thanks
Can you add more tests please? With several constraints, both passing and not passing.

**Comment by fancyweb** [comment]
I am not sure about this one because what if I want "at least 2" or "at most 3", etc. I think what you want to achieve is a composition of constraints (ref https://github.com/symfony/symfony/issues/35226), right?

When we need a composition of different constraints, shouldn't we create a custom constraint reusing core constraints "in context" instead?

What about maybe finding another name and add an option to choose the number of constraints that must be valid, so it is at least a bit dynamic?

**Comment by nicolas-grekas** [comment]
Before increasing the complexity we should be sure there is a need for the feature.

**Comment by przemyslaw-bogusz** [comment]
When I started working on this one, I thought about creating a parameter for the number of "internal" constraints, that have to be satisfied. So initially, it was the `AtLeast` constraint. But with all the discussion around color validators, I decided, that a one-of-many variant should cover 99% of cases, in which anyone would ever wanted to use such a constraint. Nevertheless, making it the `AtLeast` constraint with default option set to one, isn't such a bad idea.

By the way, I will add more tests soon.

**Comment by nicolas-grekas** [comment]
I don't think AtLeast > 1 serves any use case.

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
> Can you add more tests please? With several constraints, both passing and not passing.

File: src/Symfony/Component/Validator/Constraints/AtLeastOne.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace Symfony\Component\Validator\Constraints;
+
+/**
+ * @Annotation
+ * @Target({"PROPERTY", "METHOD", "ANNOTATION"})
+ *
+ * @author Przemysław Bogusz <przemyslaw.bogusz@tubotax.pl>
+ */
+class AtLeastOne extends Composite
```
Comment (ogizanagi) [inline src/Symfony/Component/Validator/Constraints/AtLeastOne.php]: `AtLeastOneOf` ? (would it even make sense to simplify to `Or` ?)

Comment (norkunas) [reply src/Symfony/Component/Validator/Constraints/AtLeastOne.php]: Isn't `Or` reserved keyword?

Comment (przemyslaw-bogusz) [reply src/Symfony/Component/Validator/Constraints/AtLeastOne.php]: done

**Review by OskarStark — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @przemyslaw-bogusz.

**Pull request merged**